modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/IoVersions.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.pagememory.io;
+
+import org.apache.ignite.internal.tostring.IgniteToStringBuilder;
+
+/**
+ * Registry for IO versions of the same type.
+ */
+public final class IoVersions<V extends PageIo> {
+    /**
+     * Sorted array of IO objects.
+     */
+    private final V[] vers;
+
+    /**
+     * Page type.
+     */
+    private final int type;
```

Comment:
How about creating an enum for page types, like `PageType`? It could encapsulate this `int`. The code using page types would become more readable (less confusion about 'what does this int mean'), more help from the compiler.

Reply:
I expect many independent modules to define their own IO types, enum won't work for such cases

Reply:
True. We could make `PageType` an interface (with one method `typeId()`) and add `CorePageType` that would be an enum and at the same time implement that interface.

For non-core types, we could create a class named `SimplePageType` that would just implement the `PageType` storing the id inside. Or, even better, each module would define its own `PageType`-implementing enum, like `Module1PageType`.

In such a way, we'd both have enum(s) (improving type-checking and discoverability of types in the code) and still leave a possibility for extension.